sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreDoFn.java

```diff
@@ -46,6 +46,22 @@
   @StartBundle
   public abstract void startBundle(DoFn<InT, OutT>.StartBundleContext context) throws Exception;
 
+  abstract static class NonWindowAwareDoFn<InT, OutT> extends FirestoreDoFn<InT, OutT> {
+    /**
+     * {@link ProcessContext#element() context.element()} must be non-null, otherwise a
+     * NullPointerException will be thrown.
```

Comment:
I'm not sure if this means whether Firestore Read does not support Windowing or not but please note that Windowing is a key feature of Beam and all sources are expected to support that.

Reply:
Apologies, I will need to rename this class. I'm leaning toward the new name being `ImplicitlyWindowedDoFn`, because output will exclusively use `DoFn.WindowedContext#output(OutputT)` whereas the `WindowAwareDoFn` may also output during `@FinishBundle` and must explicitly provide the window via `DoFn.FinishBundleContext#output(OutputT, org.joda.time.Instant, org.apache.beam.sdk.transforms.windowing.BoundedWindow)`